File: examples/javascript/seo_services.js
/**
 * Capilytics API - JavaScript SEO Services Examples
 */

const axios = require('axios');
const CapilyticsAuthClient = require('./authentication');

class SEOServicesClient {
  constructor(authClient) {
    this.authClient = authClient;
    this.baseUrl = authClient.baseUrl;
  }

  /**
   * Perform Google Custom Search
   */
  async googleSearch(query, options = {}) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/seo/google-search`,
        {
          query,
          max_results: options.maxResults || 10,
          language: options.language || 'en',
          date_restrict: options.dateRestrict,
          site_search: options.siteSearch,
          exclude_site: options.excludeSite
        },
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }

  /**
   * Track brand mentions
   */
  async trackBrandMentions(brandName, options = {}) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/seo/brand-mentions`,
        {
          brand_name: brandName,
          keywords: options.keywords || [],
          competitors: options.competitors || [],
          date_restrict: options.dateRestrict || 'm1',
          max_results: options.maxResults || 50
        },
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }

  /**
   * Analyze content performance
   */
  async analyzeContentPerformance(url, competitors = [], metrics = []) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/seo/content-performance`,
        {
          url,
          competitors,
          metrics
        },
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }

  /**
   * Search YouTube videos
   */
  async searchYouTubeVideos(query, options = {}) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/seo/youtube/search`,
        {
          query,
          max_results: options.maxResults || 10,
          order: options.order || 'relevance',
          published_after: options.publishedAfter,
          video_duration: options.videoDuration,
          video_type: options.videoType || 'video'
        },
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }

  /**
   * Get YouTube channel statistics
   */
  async getYouTubeChannelStats(channelId) {
    try {
      const response = await axios.get(
        `${this.baseUrl}/seo/youtube/channel/${channelId}/stats`,
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }

  /**
   * Get YouTube video analytics
   */
  async getYouTubeVideoAnalytics(videoId) {
    try {
      const response = await axios.get(
        `${this.baseUrl}/seo/youtube/video/${videoId}/analytics`,
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }

  /**
   * Track competitor YouTube videos
   */
  async trackCompetitorVideos(competitorId, channelId, options = {}) {
    try {
      const response = await axios.post(
        `${this.baseUrl}/seo/youtube/competitor/${competitorId}/videos`,
        {
          channel_id: channelId,
          max_videos: options.maxVideos || 20,
          published_after: options.publishedAfter
        },
        { headers: this.authClient.getHeaders() }
      );
      return response.data;
    } catch (error) {
      throw this.authClient.handleError(error);
    }
  }
}

// Example usage
async function main() {
  // Authenticate
  const authClient = new CapilyticsAuthClient();
  await authClient.login(
    process.env.CAPILYTICS_EMAIL || 'user@example.com',
    process.env.CAPILYTICS_PASSWORD || 'password'
  );

  const seoClient = new SEOServicesClient(authClient);

  try {
    // Example 1: Google Custom Search
    console.log('Performing Google search...');
    const searchResults = await seoClient.googleSearch(
      'competitive intelligence tools',
      {
        maxResults: 10,
        language: 'en',
        dateRestrict: 'd7'  // Last 7 days
      }
    );
    console.log(`Found ${searchResults.total_results} results`);
    console.log(`\nTop 3 results:`);
    searchResults.results.slice(0, 3).forEach((result, index) => {
      console.log(`${index + 1}. ${result.title}`);
      console.log(`   ${result.link}`);
      console.log(`   ${result.snippet}\n`);
    });

    // Example 2: Track brand mentions
    console.log('\nTracking brand mentions...');
    const mentions = await seoClient.trackBrandMentions('Capilytics', {
      keywords: ['analytics', 'competitive intelligence'],
      competitors: ['CompetitorA', 'CompetitorB'],
      dateRestrict: 'w1',  // Last week
      maxResults: 50
    });
    console.log(`Brand mentions found: ${mentions.mentions_found}`);
    console.log('Sentiment distribution:');
    Object.entries(mentions.sentiment_distribution).forEach(([sentiment, count]) => {
      console.log(`  ${sentiment}: ${count}`);
    });

    // Example 3: Analyze content performance
    console.log('\nAnalyzing content performance...');
    const performance = await seoClient.analyzeContentPerformance(
      'https://mycompany.com/blog/article',
      ['https://competitor1.com', 'https://competitor2.com'],
      ['ranking', 'backlinks', 'social_shares']
    );
    console.log(`Ranking position: ${performance.ranking_position}`);
    console.log(`Backlinks: ${performance.backlink_estimate}`);
    console.log(`Social shares: ${performance.social_shares}`);

    // Example 4: YouTube video search
    console.log('\nSearching YouTube videos...');
    const videos = await seoClient.searchYouTubeVideos(
      'competitive analysis tutorial',
      {
        maxResults: 5,
        order: 'relevance',
        videoDuration: 'medium'
      }
    );
    console.log(`Found ${videos.total_results} videos`);
    console.log('\nTop 3 videos:');
    videos.videos.slice(0, 3).forEach((video, index) => {
      console.log(`${index + 1}. ${video.title}`);
      console.log(`   Channel: ${video.channel_title}`);
      console.log(`   Published: ${video.published_at}\n`);
    });

    // Example 5: Get YouTube channel statistics
    console.log('Getting YouTube channel stats...');
    const channelId = 'UCabc123xyz';
    const channelStats = await seoClient.getYouTubeChannelStats(channelId);
    console.log(`Channel: ${channelStats.title}`);
    console.log(`Subscribers: ${channelStats.subscriber_count.toLocaleString()}`);
    console.log(`Total views: ${channelStats.view_count.toLocaleString()}`);
    console.log(`Videos: ${channelStats.video_count}`);
    console.log(`Upload frequency: ${channelStats.upload_frequency}`);

    // Example 6: Get video analytics
    console.log('\nGetting video analytics...');
    const videoId = 'abc123xyz';
    const videoAnalytics = await seoClient.getYouTubeVideoAnalytics(videoId);
    console.log(`Video: ${videoAnalytics.title}`);
    console.log(`Views: ${videoAnalytics.view_count.toLocaleString()}`);
    console.log(`Likes: ${videoAnalytics.like_count.toLocaleString()}`);
    console.log(`Comments: ${videoAnalytics.comment_count.toLocaleString()}`);
    console.log(`Engagement rate: ${videoAnalytics.engagement_rate.toFixed(2)}%`);

    // Example 7: Track competitor videos
    console.log('\nTracking competitor videos...');
    const tracking = await seoClient.trackCompetitorVideos(
      1,  // competitor_id
      'UCcompetitor123',
      {
        maxVideos: 20,
        publishedAfter: '2025-01-01T00:00:00Z'
      }
    );
    console.log(`Channel: ${tracking.channel_title}`);
    console.log(`Videos analyzed: ${tracking.videos_analyzed}`);
    console.log(`Total views: ${tracking.total_views.toLocaleString()}`);
    console.log(`Avg engagement rate: ${tracking.avg_engagement_rate.toFixed(2)}%`);
    console.log(`Upload frequency: ${tracking.upload_frequency}`);
    console.log(`\nTop topics: ${tracking.trending_topics.join(', ')}`);

  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  main();
}

module.exports = SEOServicesClient;

File: src/schemas/seo_services.py
"""Pydantic schemas for SEO Services (Google Custom Search & YouTube) API endpoints."""
from datetime import datetime
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field, validator


# Google Custom Search Schemas

class GoogleSearchRequest(BaseModel):
    """Schema for Google Custom Search request."""
    query: str = Field(..., min_length=1, description="Search query")
    max_results: int = Field(10, ge=1, le=100, description="Maximum results to return")
    language: Optional[str] = Field(None, description="Search language")
    date_restrict: Optional[str] = Field(None, description="Date restriction (d[number], w[number], m[number], y[number])")
    site_search: Optional[str] = Field(None, description="Restrict to specific site")
    exclude_site: Optional[str] = Field(None, description="Exclude specific site")


class GoogleSearchResult(BaseModel):
    """Schema for individual search result."""
    title: str
    link: str
    snippet: str
    displayed_link: Optional[str]
    pagemap: Optional[Dict[str, Any]]


class GoogleSearchResponse(BaseModel):
    """Schema for Google Custom Search response."""
    query: str
    total_results: int
    search_time: float
    results: List[GoogleSearchResult]
    next_page_token: Optional[str]


class BrandMentionRequest(BaseModel):
    """Schema for brand mention tracking request."""
    brand_name: str = Field(..., min_length=1, description="Brand name to track")
    keywords: List[str] = Field(default_factory=list, description="Additional keywords")
    competitors: List[str] = Field(default_factory=list, description="Competitor brands")
    date_restrict: Optional[str] = Field(None, description="Date restriction")
    max_results: int = Field(50, ge=1, le=100, description="Maximum results")


class BrandMentionResponse(BaseModel):
    """Schema for brand mention tracking response."""
    brand_name: str
    mentions_found: int
    sentiment_distribution: Dict[str, int]
    top_sources: List[Dict[str, Any]]
    competitor_comparison: Dict[str, int]
    trending_topics: List[str]


class ContentPerformanceRequest(BaseModel):
    """Schema for content performance analysis request."""
    url: str = Field(..., description="Content URL to analyze")
    competitors: List[str] = Field(default_factory=list, description="Competitor URLs")
    metrics: List[str] = Field(
        default=["ranking", "backlinks", "shares"],
        description="Metrics to analyze"
    )


class ContentPerformanceResponse(BaseModel):
    """Schema for content performance response."""
    url: str
    ranking_position: Optional[int]
    indexed_pages: int
    backlink_estimate: int
    social_shares: int
    competitor_comparison: Dict[str, Dict[str, Any]]
    recommendations: List[str]


# YouTube Service Schemas

class YouTubeSearchRequest(BaseModel):
    """Schema for YouTube video search request."""
    query: str = Field(..., min_length=1, description="Search query")
    max_results: int = Field(10, ge=1, le=50, description="Maximum results")
    order: str = Field("relevance", description="Sort order")
    published_after: Optional[datetime] = Field(None, description="Filter by publish date")
    video_duration: Optional[str] = Field(None, description="Video duration filter")
    video_type: Optional[str] = Field(None, description="Video type filter")

    @validator('order')
    def validate_order(cls, v):
        """Validate order parameter."""
        valid_orders = ['relevance', 'date', 'rating', 'viewCount', 'title']
        if v not in valid_orders:
            raise ValueError(f"Order must be one of: {', '.join(valid_orders)}")
        return v


class YouTubeVideoResult(BaseModel):
    """Schema for YouTube video result."""
    video_id: str
    title: str
    description: str
    channel_id: str
    channel_title: str
    published_at: datetime
    thumbnail_url: Optional[str]
    view_count: Optional[int]
    like_count: Optional[int]
    comment_count: Optional[int]


class YouTubeSearchResponse(BaseModel):
    """Schema for YouTube search response."""
    query: str
    total_results: int
    videos: List[YouTubeVideoResult]
    next_page_token: Optional[str]


class YouTubeChannelStatsRequest(BaseModel):
    """Schema for YouTube channel stats request."""
    channel_id: str = Field(..., description="YouTube channel ID")


class YouTubeChannelStatsResponse(BaseModel):
    """Schema for YouTube channel statistics."""
    channel_id: str
    title: str
    description: Optional[str]
    subscriber_count: int
    video_count: int
    view_count: int
    avg_views_per_video: float
    upload_frequency: Optional[str]
    top_videos: List[YouTubeVideoResult]


class YouTubeVideoAnalyticsRequest(BaseModel):
    """Schema for YouTube video analytics request."""
    video_id: str = Field(..., description="YouTube video ID")


class YouTubeVideoAnalyticsResponse(BaseModel):
    """Schema for YouTube video analytics."""
    video_id: str
    title: str
    channel_title: str
    published_at: datetime
    view_count: int
    like_count: int
    dislike_count: int
    comment_count: int
    favorite_count: int
    duration: str
    tags: List[str]
    category_id: str
    engagement_rate: float
    likes_to_views_ratio: float
    comments_to_views_ratio: float


class CompetitorVideoTrackingRequest(BaseModel):
    """Schema for competitor video tracking request."""
    competitor_id: int = Field(..., description="Competitor ID")
    channel_id: Optional[str] = Field(None, description="YouTube channel ID")
    max_videos: int = Field(20, ge=1, le=50, description="Maximum videos to fetch")
    published_after: Optional[datetime] = Field(None, description="Filter by publish date")


class CompetitorVideoTrackingResponse(BaseModel):
    """Schema for competitor video tracking response."""
    competitor_id: int
    channel_id: str
    channel_title: str
    videos_analyzed: int
    total_views: int
    total_likes: int
    total_comments: int
    avg_engagement_rate: float
    top_performing_videos: List[YouTubeVideoResult]
    upload_frequency: str
    trending_topics: List[str]
